### PR TITLE
Feature/alignment debug mode

### PR DESCRIPTION
**Pull request by rrahn**
Adds support for getting the alignment matrix and the trace matrix as debug configuration from the alignment result. 
Note, the actual behavior of dumping the matrix in the debug versions will be added by another PR, since I am currently refactoring a lot inside the kernel and need this functionality for debugging purposes.

### CODE REVIEW

File: include/seqan3/core/type_list/traits.hpp [hunk 1]
```diff
@@ -329,6 +329,22 @@ auto split_after(type_list<pack1_t...>, type_list<head_t, pack2_t...>)
 template <template <typename> typename trait_t, typename ...pack_t>
 type_list<trait_t<pack_t>...> transform(type_list<pack_t...>);
 
+/*!\brief Implementation for replace_at.
+ * \tparam replace_t The type replacing the old one.
+ * \tparam pack1_t   Types in the first type list.
+ * \tparam head_t    Head of the second type list.
```
Comment (joergi-w) [inline include/seqan3/core/type_list/traits.hpp]: Is this the type to be replaced? Please document so, if that's the case.

File: include/seqan3/core/type_list/traits.hpp [hunk 2]
```diff
@@ -329,6 +329,22 @@ auto split_after(type_list<pack1_t...>, type_list<head_t, pack2_t...>)
...
+ * \tparam pack2_t   Rest of the second type list.
```
Comment (joergi-w) [inline include/seqan3/core/type_list/traits.hpp]: pack1 and pack2 are collections of types before and after the element subject to replacement, right? Can you make it clear in the documentation, please?

File: include/seqan3/alignment/pairwise/align_result_selector.hpp [hunk 3]
```diff
@@ -116,8 +119,39 @@ struct align_result_selector
...
+        if constexpr (!configuration_t::template exists<align_cfg::debug>())
+        {
+            return alignment_result_value_t{};
+        }
+        else
+        {
+            using as_type_list = transfer_template_args_onto_t<alignment_result_value_t, type_list>;
+            using score_matrix_t = two_dimensional_matrix<int32_t, std::allocator<int32_t>, matrix_major_order::column>;
+            using trace_matrix_t = two_dimensional_matrix<trace_directions,
+                                                        std::allocator<trace_directions>,
```
Comment (joergi-w) [inline include/seqan3/alignment/pairwise/align_result_selector.hpp]: missing include: 
```cpp
#include <seqan3/alignment/matrix/trace_directions.hpp>
```

File: include/seqan3/alignment/pairwise/align_result_selector.hpp [hunk 4]
```diff
@@ -116,8 +119,39 @@ struct align_result_selector
...
+
+            if constexpr (configuration_t::template exists<align_cfg::result<with_alignment_type>>())
+            {
+                using with_score_t = list_traits::replace_at<score_matrix_t, 5, as_type_list>;
+                return transfer_template_args_onto_t<list_traits::replace_at<trace_matrix_t, 6, with_score_t>,
+                                                     alignment_result_value_type>{};
+            }
+            else
+            {
+                return transfer_template_args_onto_t<list_traits::replace_at<score_matrix_t, 5, as_type_list>,
```
Comment (joergi-w) [inline include/seqan3/alignment/pairwise/align_result_selector.hpp]: When I set the debug config, I receive always the score matrix, and only if the alignment was requested I additionally receive the trace matrix. Please describe this behaviour in the documentation.

File: include/seqan3/alignment/pairwise/alignment_result.hpp [hunk 5]
```diff
@@ -200,6 +209,44 @@ class alignment_result
...
+    //!\cond DEV
+    /*!\brief Returns the score matrix used to compute the alignment.
+     * \return The score matrix.
+     *
+     * \details
+     *
+     * This function is only used for debugging such that performance can be affected significantly when enabling
+     * seqan3::align_cfg::debug.
+     *
+     * \attention This function with fail the compilation, if the alignment was not requested in the alignment
```
Comment (joergi-w) [inline include/seqan3/alignment/pairwise/alignment_result.hpp]: Inconsistency: In the selector in `augment_if_debug` you make a distinction about whether alignment has been selected only for the trace matrix (and not for the score matrix). And here is stated that alignment is necessary for the score matrix, which is not true for your implementation!

File: include/seqan3/alignment/pairwise/alignment_result.hpp [hunk 6]
```diff
@@ -200,6 +209,44 @@ class alignment_result
...
+
+    /*!\brief Returns the trace matrix used to compute the alignment.
+     * \return The score matrix.
+     *
+     * \details
+     *
+     * This function is only used for debugging such that performance can be affected significantly when enabling
+     * seqan3::align_cfg::debug.
+     *
+     * \attention This function with fail the compilation, if the alignment was not requested in the alignment
```
Comment (joergi-w) [inline include/seqan3/alignment/pairwise/alignment_result.hpp]: Shouldn't you add that also the _debug_ config is required?

File: include/seqan3/alignment/pairwise/align_result_selector.hpp (hunk 4 above)
Comment (rrahn) [reply include/seqan3/alignment/pairwise/align_result_selector.hpp]: It is?
> Using this configuration allows to output the alignment score matrix and if requested the alignment trace matrix.\

I rephrase it to make it more clear.

File: include/seqan3/alignment/pairwise/alignment_result.hpp (hunk 6 above)
Comment (rrahn) [reply include/seqan3/alignment/pairwise/alignment_result.hpp]: Well, probably yes. It is only visible to developers though.

File: include/seqan3/alignment/configuration/align_config_debug.hpp [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+ * returned seqan3::alignment_result.
+ * The score matrix is always accessible, while the trace matrix can only be computed if an alignment was
+ * requested via the seqan3::align_cfg::result configuration.
+ *
+ * \note This configuration is only useful for debugging purposes as it can have a significant impact on the
+ *       performance.
+ */
+inline constexpr debug_type debug{};
+//!\endcond
+}  // namespace seqan3::align_cfg
```
Comment (h-2) [inline include/seqan3/alignment/configuration/align_config_debug.hpp]: why not ::detail?

File: include/seqan3/alignment/pairwise/align_result_selector.hpp [hunk 8]
```diff
@@ -116,8 +120,39 @@ struct align_result_selector
         }
     }
 
-    //!\brief The determined result type.
-    using type = decltype(_determine());
+    //!\brief Augments the result type with the debug output for score and trace matrix.
+    template <typename alignment_result_value_t>
+    static constexpr auto augment_if_debug(alignment_result_value_t)
+    {
+        if constexpr (!configuration_t::template exists<align_cfg::debug_type>())
```
Comment (h-2) [inline include/seqan3/alignment/pairwise/align_result_selector.hpp]: In general, it is better style to not negate the condition and instead just switch the code blocks, but that's :nail_care:

File: include/seqan3/alignment/pairwise/alignment_result.hpp [hunk 9]
```diff
@@ -19,17 +19,21 @@ namespace seqan3::detail
...
  */
 template <typename id_t,
           typename score_t,
           typename back_coord_t = std::nullopt_t *,
           typename front_coord_t = std::nullopt_t *,
-          typename alignment_t = std::nullopt_t *>
+          typename alignment_t = std::nullopt_t *,
+          typename score_debug_matrix_t = std::nullopt_t *,
+          typename trace_debug_matrix_t = std::nullopt_t *>
 struct alignment_result_value_type
```
Comment (h-2) [inline include/seqan3/alignment/pairwise/alignment_result.hpp]: Why is the alignment_result type in detail? Aren't people to supposed to know how to access it?

File: include/seqan3/alignment/pairwise/alignment_result.hpp [hunk 10]
```diff
@@ -200,6 +209,43 @@ class alignment_result
         return data.alignment;
     }
     //!\}
+
+    //!\cond DEV
```
Comment (h-2) [inline include/seqan3/alignment/pairwise/alignment_result.hpp]: Why is this wrapped in DEV when the class is part of detail anyway?

File: include/seqan3/core/detail/empty_type.hpp [hunk 11]
```diff
@@ -0,0 +1,25 @@
...
+#pragma once
+
+#include <seqan3/core/platform.hpp>
+
+namespace seqan3::detail
+{
+
+//!\brief An empty class type used in meta programming.
+//!\ingroup core
+struct empty_type
```
Comment (h-2) [inline include/seqan3/core/detail/empty_type.hpp]: I think we have this on other places, too, maybe move to core?

Comment (h-2) [reply include/seqan3/core/detail/empty_type.hpp]: Oh, it is in code already. Ok. One place that you need to change is inherited_iterator_base, because that exactly also defines this type (with the same name).

File: include/seqan3/alignment/configuration/align_config_debug.hpp (hunk 7 above)
Comment (rrahn) [reply include/seqan3/alignment/configuration/align_config_debug.hpp]: Mhmm, yes, I could put the config in detail.

Comment (rrahn) [reply include/seqan3/alignment/configuration/align_config_debug.hpp]: For now it is only visible for developer but I found it nicer to have the chain use `align_cfg` as well. But this is not so important and detail would be fine as well

File: include/seqan3/alignment/pairwise/alignment_result.hpp (hunk 9 above)
Comment (rrahn) [reply include/seqan3/alignment/pairwise/alignment_result.hpp]: That is a internal value type used to store all the members and to selectively disable some of the values in the template argument. This way the user does not need to bother with the template arguments and only sees a simple class that has member functions documented properly. Of course the entire type signature is therefor implementation defined and this is not ideal but for the moment the best compromise I could find.

File: include/seqan3/core/detail/empty_type.hpp (hunk 11 above)
Comment (rrahn) [reply include/seqan3/core/detail/empty_type.hpp]: It was moved from there

File: include/seqan3/alignment/pairwise/alignment_result.hpp (hunk 10 above)
Comment (rrahn) [reply include/seqan3/alignment/pairwise/alignment_result.hpp]: No the class itself is not in detail. Only the internal value type.

**Comment by rrahn** [comment]
> On a general design level: Do you expect more debug stuff? Would it make sense to have a extra debug member of the return type and put the other stuff inside so all debug things are wrapped together? I am sure the index could use something similar...

I do not expect more debug stuff for now. But something we can keep in mind for later point. The entire debug stuff is not visible to the user anyway

File: include/seqan3/alignment/configuration/align_config_debug.hpp (hunk 7 above)
Comment (rrahn) [reply include/seqan3/alignment/configuration/align_config_debug.hpp]: So I changed it to some global config thing. But note that the actual debug thing need to be in align_cfg to disambiguate it from other definitions in other namespaces. Like for the search you need your own config element since it uses internal enum ids to validate the combinability. Thus leaving the actual variable in align_cfg makes sense. But I put the global type into detail and to some generic place usable for other algorithms

Comment (h-2) [reply include/seqan3/alignment/configuration/align_config_debug.hpp]: ok

File: include/seqan3/alignment/pairwise/alignment_result.hpp (hunk 9 above)
Comment (h-2) [reply include/seqan3/alignment/pairwise/alignment_result.hpp]: Ok, I don't fully know how it works, I just wanted to bring your attention to it.

File: include/seqan3/core/detail/empty_type.hpp (hunk 11 above)
Comment (h-2) [reply include/seqan3/core/detail/empty_type.hpp]: Hm, I didn't see it being removed there, but maybe it was.

File: include/seqan3/alignment/pairwise/alignment_result.hpp [hunk 12]
```diff
@@ -190,7 +199,7 @@ class alignment_result
 
     /*!\brief Returns the actual alignment, i.e. the base pair matching.
      * \return At least two aligned sequences, which represent the alignment.
-     * \attention This function with fail the compilation, if the alignment was not requested in the alignment
+     * \attention Compiling this function will fail, if the alignment was not requested in the alignment
```
Comment (h-2) [inline include/seqan3/alignment/pairwise/alignment_result.hpp]: As mentioned in the other PR, I don't fully get this, but we can discuss it after it has been merged.
